[FLINK-21439][core] WIP: Adds Exception History for AdaptiveScheduler

bytesandwich opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

This pull request will address saving exception histories in `AdaptiveScheduler`.



## Brief change log
- three tests in `aptiveSchedulerTest` for 3 conditions: knowing the specific failing task, for not knowing, and for concurrent failures.
- added the history queue to `AdaptiveScheduler` along with `requestJob`
- added `archiveExecutionFailure` to `StateWithExecutionGraph` to populate the queue in `AdaptiveScheduler` along with uses of this method in `archiveAnyFailure` in `Executing` and `StopWithSavepoint`.
- added `failingExecutionVertexId` to `Executing.FailureResult`.
- added `userCodeClassLoader` to `StateWithExecutionGraph`
-  add error archiving to `Failing` and `Restarting`
- added a `FailureHandlingResultSnapshot` factory method that accepts what we have available in AdaptiveScheduler when we have failures. I think maybe it's a broader type of FailureSnapshot now. 


## Verifying this change

This change adds exception state accumulation to the `AdaptiveScheduler` and has associated tests in `AdaptiveSchedulerTest`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): noi
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: (yes / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveScheduler.java:
```diff
@@ -145,14 +151,14 @@
-                Failing.Context,
-                Finished.Context,
-                StopWithSavepoint.Context {
+        Created.Context,
+        WaitingForResources.Context,
+        CreatingExecutionGraph.Context,
+        Executing.Context,
+        Restarting.Context,
+        Failing.Context,
+        Finished.Context,
+        StopWithSavepoint.Context {
```
Hi @bytesandwich , could you update your IDE's project settings following the [code formatting setup recommended in the Flink docs|https://ci.apache.org/projects/flink/flink-docs-master/docs/flinkdev/ide_setup/#code-formatting]? That would avoid changes like that to happen. I'm gonna have a look at the PR after you updated it.

bytesandwich replied:
hi @XComp thanks for pointing that out. my google-java-format plugin was enabled in the plugin install page but was disabled in its own settings, so going through the steps fixed it and removed a lot of noise.

Here are some things you might also consider in your review:

- Maybe Executing and StopWithSavepoint shouldn't always call archiveExecutionFailure as I have in the PR?

- Should we rename the class `FailureHandlingResultSnapshot` to `FailureSnapshot` or something like that?

Will this work? @flinkbot run travis

bytesandwich replied:
ping @XComp

XComp replied:
Sorry for the late response. I am still busy with other stuff. I hope to get back to you today or on Monday.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/ExecutingTest.java:
```diff
@@ -74,8 +74,8 @@
 import java.util.concurrent.ScheduledFuture;
+import java.util.function.BiFunction;
 import java.util.function.Consumer;
-import java.util.function.Function;
```
`Function` is still used in this test class (in `MockExecutionJobVertex`) and shouldn't be removed.

bytesandwich replied:
Done

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerTest.java:
```diff
@@ -792,11 +801,140 @@ public void testHowToHandleFailureUnrecoverableFailure() throws Exception {
+        assertThat(
+                failure.getException().deserializeError(ClassLoader.getSystemClassLoader()),
+                Matchers.is(expectedException));
+        assertThat(failure.getTimestamp(), greaterThanOrEqualTo(start));
+        assertThat(failure.getTimestamp(), lessThanOrEqualTo(end));
+        assertThat(failure.getTaskManagerLocation(), Matchers.is(nullValue()));
+        assertThat(failure.getFailingTaskName(), Matchers.is(nullValue()));
```
There's a `ExceptionHistoryEntryMatcher` which you could use (and extend) instead. @

bytesandwich replied:
Do we need to add the code from `getFailureTimestamp` in `DefaultSchedulerTest` to use that?

XComp reviewed (changes requested):
Thanks @bytesandwich for your draft. And thanks for your patience. I looked through your proposal and the changes look good already. I had a few remarks and questions which are listed below. Looking forward to your reply.

bytesandwich commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java:
```diff
@@ -59,27 +59,49 @@
+    }
+
+    /**
+     * Creates a {@code FailureHandlingResultSnapshot} based on the passed {@link
+     * FailureHandlingResult} and {@link ExecutionVertex ExecutionVertices}.
```
Done

XComp replied:
I was referring to that bit:
```suggestion
     * Creates a {@code FailureHandlingResultSnapshot} based on the passed {@link
     * FailureHandlingResult} and {@link ExecutionVertex ExecutionVertices}.
```
The description does not match the method's signature.

bytesandwich commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveScheduler.java:
```diff
@@ -880,6 +891,19 @@ public void goToFinished(ArchivedExecutionGraph archivedExecutionGraph) {
+    }
+
+    private Iterable<RootExceptionHistoryEntry> getExceptionHistory() {
+        final Collection<RootExceptionHistoryEntry> copy = new ArrayList<>(exceptionHistory.size());
+        exceptionHistory.forEach(copy::add);
+        return copy;
```
Does making {BoundedFIFOQueue} an  [AbstractQueue](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/AbstractQueue.html) and replacing this with {new ArrayList{<the queue>)} work?

bytesandwich commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java:
```diff
@@ -122,6 +132,15 @@ boolean updateTaskExecutionState(TaskExecutionStateTransition taskExecutionState
     }
 
+    @Nullable
+    private ExecutionVertexID getExcutionVertexId(ExecutionAttemptID id) {
```
Done

bytesandwich commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerTest.java:
```diff
@@ -763,7 +771,8 @@ public void testHowToHandleFailureRejectedByStrategy() throws Exception {
 
-        assertThat(scheduler.howToHandleFailure(new Exception("test")).canRestart(), is(false));
+        assertThat(
+                scheduler.howToHandleFailure(null, new Exception("test")).canRestart(), is(false));
```
Done

bytesandwich commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java:
```diff
@@ -293,9 +316,19 @@ void goToFailing(
     static final class FailureResult {
```
I think there's a follow up refactor of `handleAnyFailure` etc to be had.

bytesandwich commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java:
```diff
@@ -122,6 +132,15 @@ boolean updateTaskExecutionState(TaskExecutionStateTransition taskExecutionState
         return successfulUpdate;
     }
 
+    @Nullable
+    private ExecutionVertexID getExcutionVertexId(ExecutionAttemptID id) {
+        Execution execution = getExecutionGraph().getRegisteredExecutions().get(id);
+        if (execution == null) {
+            return null;
+        }
+        return execution.getVertex().getID();
+    }
+
```
Done

bytesandwich commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerTest.java:
```diff
@@ -792,11 +801,140 @@ public void testHowToHandleFailureUnrecoverableFailure() throws Exception {
+
+        final Exception expectedException = new Exception("Expected Global Exception");
+        final long start = System.currentTimeMillis();
+        final CountDownLatch latch = new CountDownLatch(1);
```
Done.

bytesandwich commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepoint.java:
```diff
@@ -178,9 +192,13 @@ private void completeOperationAndGoToFinished(String savepoint) {
     }
 
-    private void handleAnyFailure(Throwable cause) {
+    private void handleAnyFailure(
```
I agree that this copied code smells really bad. *I think for this change we should limit the scope since it's already been more than a month and it's a minor feature about exposing exceptions.*

For architecting failure handling, I think if the specifics start to be challenging, then it's good to step back and take a long-term outlook of what the architecture should evolve into. We have at least three examples to consider. The [AdaptiveScheduler FLIP](https://cwiki.apache.org/confluence/display/FLINK/FLIP-160%3A+Adaptive+Scheduler#FLIP160:AdaptiveScheduler-Stablesetofresources) refers to [rebalancing in kafka](https://www.confluent.io/blog/cooperative-rebalancing-in-kafka-streams-consumer-ksqldb/). I believe beam (and Google Dataflow) use a [splitting](https://beam.apache.org/blog/splitatfraction-method/) abstraction to rebalance and tolerate live failures. And, the `DefaultScheduler` has fine-grained recovery as you know. So, how could we design to "plug/unplug" all or any of these into `AdaptiveScheduler` and keep the scheduler as simple/dumb as possible? That's a pretty tough question but a simple answer to that question would probably create a good architecture.

Maybe we could make all these things simpler by first extracting _all_ the duplicated code that accesses and retrieves things from the execution graph into some `GraphUtil` or something like that. E.g. 
```
        Set<ExecutionVertexID> concurrentVertexIds =
                IterableUtils.toStream(getExecutionGraph().getSchedulingTopology().getVertices())
                        .map(SchedulingExecutionVertex::getId)
                        .filter(
                                v ->
                                        failingExecutionVertexId != null
                                                && !failingExecutionVertexId.equals(v))
                        .collect(Collectors.toSet());
```
in `StateWithExecutionGraph` -> `archiveExecutionFailure` because this is so much deep coupling to the execution graph! 

Maybe something like a declarative interface that sits on top of a visitor layer that sits on top of the actual graph data?

XComp replied:
Thanks for sharing your thoughts. You're right: Tackling the failure handling for the `AdaptiveScheduler` might be a bit out of scope of this ticket. There are some next steps that include making the failure handling of the `AdaptiveScheduler` more sophisticated. That might be the better place to fix that. Let's focus on the exception history for now.

XComp commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerTest.java:
```diff
@@ -792,11 +809,140 @@ public void testHowToHandleFailureUnrecoverableFailure() throws Exception {
+        assertThat(failure.getTaskManagerLocation(), Matchers.is(nullValue()));
+        assertThat(failure.getFailingTaskName(), Matchers.is(nullValue()));
+    }
+
```
Can you add the concurrent failure test here as well? This should fail right now since we're not covering the failure archiving in the restart state.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java:
```diff
@@ -81,11 +83,15 @@ public void cancel() {
+        final FailureResult failureResult =
+                context.howToHandleFailure(failingExecutionVertexId, cause);
+
+        archiveExecutionFailure(failingExecutionVertexId, cause);
```
Thinking about it once more: I guess, it's not the right location to archive the failure considering that we also want to identify concurrent failures. We haven't addressed that in this PR, yet.

To achieve that, we have to collect the failure snapshot here and pass it over to the next state (failure or restart). Any failure that pops up in these subsequent states has to be collected as well. The archiving should happen when re-instantiating the `ExecutionGraph`.

XComp replied:
We should also cover this in the corresponding `StateWithExecutionGraphTest` test implementations.

XComp reviewed (changes requested):
Thanks @bytesandwich . The changes go into the right direction, I think. 👍  We're still missing the support of concurrent failures. I tried to pin-point this in some of my comments below. Looking forward to your response.

XComp commented:
@bytesandwich just letting you know: The feature freeze for Flink 1.14 is beginning of August 2021. How realistic is it finish this PR till then? FYI: @zentol might take over as I'm out for the next weeks.

bytesandwich commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StateWithExecutionGraph.java:
```diff
@@ -110,6 +121,22 @@ ExecutionGraph getExecutionGraph() {
+    @Nullable
+    protected ExecutionVertexID getExecutionVertexId(ExecutionAttemptID id) {
+        Execution execution = getExecutionGraph().getRegisteredExecutions().get(id);
+        if (execution == null) {
```
Is this the same behavior as the original method: https://github.com/apache/flink/blob/d105eb32cda5f903eebc7a1019de44fdc1ec8472/flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java#L519 ?

XComp replied:
Not sure whether I understand you correctly: The method implementation exists in `SchedulerBase` as well, yes.

zentol commented:
@bytesandwich ping

bytesandwich commented:
Hi @zentol I just got back from a vacation so I'm looking at this again. I'm not sure what we want specification wise with concurrent failure support. I can imagine all sorts of things failing concurrently. It seems like maybe concurrent failures would be best tested in a more elaborate integration test to have a very clear expectation of correct behavior? Perhaps that could be a follow up ticket to make this current minimal exception handling landable quickly after I address the current feedback?

bytesandwich commented:
Regarding using `ExceptionHistoryEntryMatcher`, I'm not sure what exact timestamp to expect. Maybe it's best to stick with the range that's in the test now without the matcher?

FYI I uploaded the changes, I had made before vacation, that I think addressed the review. I'd like to see what the integration test runs do, since developing on windows makes it hard to run the build.

XComp commented:
> Hi @zentol I just got back from a vacation so I'm looking at this again. I'm not sure what we want specification wise with concurrent failure support. I can imagine all sorts of things failing concurrently. It seems like maybe concurrent failures would be best tested in a more elaborate integration test to have a very clear expectation of correct behavior? Perhaps that could be a follow up ticket to make this current minimal exception handling landable quickly after I address the current feedback?

Hi @bytesandwich , I'm back from vacation so I am able to answer your questions.
Testing concurrent failures should be possible as part of the `AdaptiveSchedulerTest`. Similarly to what you've done in [AdaptiveSchedulerTest:929](https://github.com/apache/flink/blob/521e19eecadf39a226c5b5be4ed5348485656eab/flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerTest.java#L929) with one `updateTaskExecutionState` call you should be able to do with two calls. The first call will make the `AdaptiveScheduler` switch into restarting state. Calling the `updateTaskExecutionState` again would not catch the second exception right now. Implementing the exception handling also in the `Restarting` state class should solve the issue.

Analogously, that has to be done for cases where the scheduler does not switch into `Restarting` state but `Failing`. Does that make sense to you?

XComp commented:
> Regarding using `ExceptionHistoryEntryMatcher`, I'm not sure what exact timestamp to expect. Maybe it's best to stick with the range that's in the test now without the matcher?
> 
> FYI I uploaded the changes, I had made before vacation, that I think addressed the review. I'd like to see what the integration test runs do, since developing on windows makes it hard to run the build.

The same way, how you get the `ExecutionAttemptID` in [AdaptiveSchedulerTest:920-924](https://github.com/apache/flink/blob/521e19eecadf39a226c5b5be4ed5348485656eab/flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveSchedulerTest.java#L920-L924) you could also get the failure info. Just instead of calling `getAttemptId()` to retrieve the ID, you could call `getFailureInfo()` to retrieve the `ErrorInfo` which includes the timestamp after `updateTaskExecutionState` is called.

I hope that helped.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java:
```diff
@@ -59,38 +58,60 @@
+     * @return The {@code FailureHandlingResultSnapshot}.
+     */
+    public static FailureHandlingResultSnapshot create(
+            ExecutionVertexID failingExecutionVertexId,
```
```suggestion
            @Nullable ExecutionVertexID failingExecutionVertexId,
```

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/exceptionhistory/FailureHandlingResultSnapshot.java:
```diff
@@ -59,38 +58,60 @@
+    /**
+     * Creates a {@code FailureHandlingResultSnapshot} based on passed parameters.
+     *
+     * @param failingExecutionVertexId an {@link Optional} of the {@link ExecutionVertexID} the
+     *     failure originates from, or {@code None}.
```
The JavaDoc would need to be updated as well since it's a `@Nullable` object instead of an `Optional` now.

XComp reviewed (changes requested):
Thanks @bytesandwich. I looked through your questions and changes and responded to all of them. Please see my comments above and below. Feel free to reach out for further questions.

bytesandwich commented:
Hi @XComp I just switched jobs so I was out for a bit. I see how you envision the test and I implemented it that way. PTAL!

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Canceling.java:
```diff
@@ -37,8 +37,15 @@
             OperatorCoordinatorHandler operatorCoordinatorHandler,
-            Logger logger) {
-        super(context, executionGraph, executionGraphHandler, operatorCoordinatorHandler, logger);
+            Logger logger,
```
Just for the record: I verified that we do not collect failures while being already in cancelling stage for the `DefaultScheduler` (see [source](https://github.com/apache/flink/blob/35721033e18da6cb3e552de637942664b1ad879d/flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java#L1111)). Hence, not implementing the archiving in this class is correct. 👍

XComp reviewed (changes requested):
Thanks, @bytesandwich . The changes look good. We're going into the right direction with it. I added some comments to your code changes. Please check them below.

I didn't go through the tests, yet. AzureCI seems to be failing quite a bit. Is this related to your changes?

XComp commented:
@bytesandwich Any update on your side? If not, I might pick it up to finalize it for Flink 1.15 after the upcoming release of 1.14.

bytesandwich commented:
Hi @XComp I think that would be a good idea, unless you feel that the changes you asked for are the final changes to land the diff? I am not working with Flink at my day job at the moment.

XComp commented:
This PR was superseded by https://github.com/apache/flink/pull/18689. Closing the PR.
